scripts/validate.js

```diff
@@ -0,0 +1,66 @@
+import path from 'path'
+import Ajv from 'ajv'
+import { readdirSync, readFileSync } from 'fs'
+
+const ajv = new Ajv()
+
+const SAMPLES_DIRECTORY = './samples'
+const SCHEMAS_DIRECTORY = './schemas'
+
+forEachFile(SCHEMAS_DIRECTORY, addSchema)
+forEachFile(SAMPLES_DIRECTORY, validateSample)
+
+function addSchema(schemaPath) {
+  const schema = readJson(schemaPath)
+
+  // We need to be careful about schema ids because they need to:
+  // * be unique, or else Ajv will throw
+  // * represent a path, as Ajv will use it to resolve $refs
+  // Here, we make sure that both requirements are respected.
+  const schemaId = computeSchemaIdFromSchemaPath(schemaPath)
+  if (schema.$id !== schemaId) {
+    console.log(`⚠️ Schema ${schemaPath} $id should be ${schemaId}`)
+    process.exitCode = 1
+  }
+
+  ajv.addSchema(schema)
+}
+
+function validateSample(samplePath) {
+  const schemaId = computeSchemaIdFromSamplePath(samplePath)
+  const valid = ajv.validate(schemaId, readJson(samplePath))
+  if (valid) {
+    console.log(`✅ ${samplePath}`)
+  } else {
+    console.log(`❌ ${samplePath} is not valid against ${schemaId}:`)
+    console.log(`   - ${ajv.errorsText(undefined, { separator: '\n   - ' })}`)
+    process.exitCode = 1
+  }
+}
+
+function computeSchemaIdFromSchemaPath(schemaPath) {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it’s relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** Does it make sense to keep file extension `.json` for IDs?

Reply:
So I tried and technically it is possible to remove the extension from IDs, but because schemas `$ref` are resolved using IDs, we need to change `$ref` as well. It works with `ajv`, but other json-schema libraries might resolve `$ref` using file paths, so `$ref` should still represent a path name.

To keep compatibility with other libraries, let's keep using a real file path.